docs/design/settings/extensions/modular-settings-extensions.md

```diff
@@ -0,0 +1,314 @@
+Any one settings extension has the power to halt the transaction by returning a non-zero exit-code on validation.
+
+Validations must specify the setting which triggers the validation, as well as a list of templates which will be rendered and provided to the validator.
+
+```toml
+# auditor.toml -- setting extension config file for `settings.auditor`
+[[extension.validations]]
+validates = "network.hostname"
+with-templates = [
+    "/templates/auditor-extension/validator-template.toml"
+]
+```
+
+When a validation is triggered, it gains access to the new state via rendered templates.
+The validator is also provided the name of the setting which was changed, as well as access to its own settings data that it can use as configuration.
+
+#### Datastore Layout
+Once the `apiserver` has a target set of data to write for a given `set` request and it has been validated by all required validators, it must persist the new settings to the datastore.
+Much like the current datastore, stored settings are stored to the filesystem via a particular pattern.
+
+```bash
+datastore/
+└── motd
+    ├── v1
+    │   └── motd.json
+    └── v2
+        └── motd.json
+```
+Any data written to directories created during a `set` transaction is moved into the datastore.
+Despite only one version of the data being provided, the `apiserver` will populate all versions supported by the settings extension at this time by requesting that the settings extension perform all necessary migrations.
+Failure to populate a version listed as "supported" will result in a failed transaction.
+See the next section about migrations for more details.
+
+### Extension Versioning and Migrations
+Bottlerocket's settings data model changes frequently: sometimes to support new features, or sometimes to correct mistaken model shapes.
+In order to support changes to the settings model, settings must be capable of exposing versioning information about the settings extension, as well as providing faculties for migrating settings data to new versions.
+Our versioning scheme will take heavy influence from Kubernetes' scheme for [versioning CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/).
+Settings extension versioning information is surfaced through the extension config file:
+
+```toml
+# hostname.toml -- setting extension config file for settings.hostname
+[extension]
+name = "hostname",
+supported-versions = [
+     "v1",
+     "v2"
+]
+default-version = "v2"
+```
+
+Any setting can be written or read at any of the `supported-versions` listed by the extension.
+Unversioned requests to read or write this setting are assumed to be the `default-version`.
+During a `set` transaction, the `apiserver` will consider the provided version to be the `canonical-version`, and will ask the settings extension to perform migrations from the `canonical` version to all supported versions using the `migrate` command.
```

Comment:
Should `migrate` be rolled into `set`? i.e. should `set` also do the "migrations" since there is no case when `set` would *not* be followed by `migrate`?